swag/assistant.py

```diff
@@ -1,32 +1,137 @@
+from typing import List, Dict, Any, Optional
+from pathlib import Path
 from anthropic import Anthropic
-from anthropic.types import Message
-import os
-from dotenv import load_dotenv
-from typing import Any
+from anthropic.types import TextBlock, ToolUseBlock
+import base64
+from swags.tools import ToolRegistry
 
-load_dotenv()
 
-def query_claude(messages: list, system: str, tools: list) -> Message:
+class Assistant:
+    def __init__(
+        self,
+        client: Anthropic,
+        model: str,
+        system: Optional[str] = None,
+        tools: List[Dict] = [],
+        max_steps: int = 10,
+    ):
+        self.client = client
+        self.model = model
+        self.system = system
+        self.max_steps = max_steps
+        self.messages = []
+        self.max_tokens = 1024
+        self.define_tools(tools)
 
-    client = Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
-    response = client.messages.create(
-            model="claude-3-5-haiku-latest",
-            max_tokens=1024,
-            system=system,
-            messages=messages,
-            tools=tools
-    )
+    def define_tools(self, tools: List[type]):
+        """Defines the tools available to the assistant."""
+        if tools:
+            self.tools = [convert_pydantic_to_anthropic_schema(tool) for tool in tools]
+            self.tool_fns = {
+                tool.__name__: ToolRegistry.get(tool.__name__)[0] for tool in tools
+            }
+        else:
+            self.tools = []
+            self.tool_fns = {}
+
+    def __call__(
+        self,
+        prompt: str,
+        image: Optional[str] = None,
+        current_result: Any = None,
+    ) -> Optional[str]:
+        if len(self.messages) >= 2 * self.max_steps:
+            print("Maximum number of steps reached.")
+            return current_result
+
+        message = {"role": "user", "content": [{"type": "text", "text": prompt}]}
+        self.messages.append(message)
+
+        if image:
+            image_string = self.load_image_base64(Path(image))
+            message["content"].append(
+                {
+                    "type": "image",
+                    "source": {
+                        "type": "base64",
+                        "media_type": "image/jpeg",
+                        "data": image_string,
+                    },
+                }
+            )
+
+        try:
+            response = self.client.messages.create(
+                model=self.model,
+                max_tokens=self.max_tokens,
+                tools=self.tools or [],
+                system=self.system or "",
+                messages=self.messages,
+            )
+
+            assistant_message = {"role": "assistant", "content": []}
+            for content in response.content:
+                if isinstance(content, TextBlock):
```

Comment:
We need to yield the response back so we can stream it.

Reply:
Updated. when invoking the endpoint, the block content (object) is streamed